components/raftstore/src/store/worker/compact.rs

```diff
@@ -120,33 +210,83 @@ pub enum Error {
 }
 
 impl<E> Runner<E>
 where
     E: KvEngine,
 {
-    pub fn new(engine: E) -> Runner<E> {
-        Runner { engine }
+    pub fn new(engine: E, remote: Remote<yatp::task::future::TaskCell>) -> Runner<E> {
+        Runner { engine, remote }
     }
 
     /// Periodic full compaction.
+    /// Note: this does not accept a `&self` due to async lifetime issues.
     ///
-    /// NOTE this is a highly experimental feature!
+    /// NOTE this is an experimental feature!
     ///
-    /// TODO: Do not start if there is heavy I/O.
-    /// TODO: Make it possible to rate limit, pause, or abort this by compacting
-    /// a range at a time.
-    pub fn full_compact(&mut self) -> Result<(), Error> {
+    /// TODO: Support stopping a full compaction.
+    async fn full_compact(
+        engine: E,
+        ranges: Vec<(Key, Key)>,
+        compact_controller: FullCompactController,
+    ) -> Result<(), Error> {
         fail_point!("on_full_compact");
         info!("full compaction started");
+        let mut ranges: VecDeque<_> = ranges
+            .iter()
+            .map(|(start, end)| (Some(start.as_slice()), Some(end.as_slice())))
+            .collect();
+        if ranges.is_empty() {
+            ranges.push_front((None, None))
+        }
+
         let timer = Instant::now();
         let full_compact_timer = FULL_COMPACT.start_coarse_timer();
-        box_try!(self.engine.compact_range(
-            None, None, // Compact the entire key range.
-            true, // no other compaction will run when this is running
-            1,    // number of threads threads
-        ));
+
+        while let Some(range) = ranges.pop_front() {
+            debug!(
+                "incremental range full compaction started";
+            "start_key" => ?range.0.map(log_wrappers::Value::key),
+            "end_key" => ?range.1.map(log_wrappers::Value::key),
+             );
+            let incremental_timer = FULL_COMPACT_INCREMENTAL.start_coarse_timer();
+            box_try!(engine.compact_range(
+                range.0, range.1, // Compact the entire key range.
+                false,   // non-exclusive
+                1,       // number of threads threads
+            ));
+            incremental_timer.observe_duration();
+            debug!(
+                "finished incremental range full compaction";
+                "remaining" => ranges.len(),
+            );
+            // If there is at least one range remaining in `ranges` remaining, evaluate
+            // `compact_controller.incremental_compaction_pred`. If `true`, proceed to next
+            // range; otherwise, pause this task
+            // (see `FullCompactController::pause` for details) until predicate
+            // evaluates to true.
+            if let Some(next_range) = ranges.front() {
+                if !(compact_controller.incremental_compaction_pred)() {
+                    warn!("pausing full compaction before next increment";
```

Comment:
use info! is more reasonable to me as here the pausing is by design.

Reply:
done.